MapboxMobileEvents/MMEEvent.m

```diff
@@ -3,206 +3,192 @@
 }
 
+#pragma mark - Custom Events
+
 + (instancetype)turnstileEventWithAttributes:(NSDictionary *)attributes {
-    MMEEvent *turnstileEvent = [[MMEEvent alloc] init];
-    turnstileEvent.name = MMEEventTypeAppUserTurnstile;
-    turnstileEvent.attributes = attributes;
-    return turnstileEvent;
+    NSMutableDictionary *eventAttributes = attributes.mutableCopy;
+    eventAttributes[MMEEventKeyEvent] = MMEEventTypeAppUserTurnstile;
+
+    return [MMEEvent eventWithAttributes:eventAttributes];
+}
+
++ (instancetype)visitEventWithAttributes:(NSDictionary *)attributes {
+    NSMutableDictionary *eventAttributes = attributes.mutableCopy;
+    eventAttributes[MMEEventKeyEvent] = MMEEventTypeVisit;
+
+    return [MMEEvent eventWithAttributes:eventAttributes];
+}
+
++ (instancetype)debugEventWithAttributes:(NSDictionary *)attributes {
+    NSMutableDictionary *eventAttributes = attributes.mutableCopy;
+    eventAttributes[MMEEventKeyEvent] = MMEEventTypeLocalDebug;
+
+    return [MMEEvent eventWithAttributes:eventAttributes];
+}
+
++ (instancetype)debugEventWithError:(NSError*) error {
+    NSMutableDictionary* eventAttributes = [NSMutableDictionary dictionaryWithObject:MMEDebugEventTypeError forKey:MMEDebugEventType];
+    eventAttributes[MMEEventKeyErrorCode] = @(error.code);
+    eventAttributes[MMEEventKeyErrorDescription] = (error.localizedDescription ? error.localizedDescription : error.description);
+    eventAttributes[MMEEventKeyErrorFailureReason] = (error.localizedFailureReason ? error.localizedFailureReason : MMEEventKeyErrorNoReason);
+
+    return [self debugEventWithAttributes:eventAttributes];
+}
+
++ (instancetype)debugEventWithException:(NSException*) except {
+    NSMutableDictionary* eventAttributes = [NSMutableDictionary dictionaryWithObject:MMEDebugEventTypeError forKey:MMEDebugEventType];
+    eventAttributes[MMEEventKeyErrorDescription] = except.name;
+    eventAttributes[MMEEventKeyErrorFailureReason] = (except.reason ? except.reason : MMEEventKeyErrorNoReason);
+
+    return [self debugEventWithAttributes:eventAttributes];
+}
+
+#pragma mark - Deperecated
+
++ (instancetype)eventWithDate:(NSDate *)eventDate name:(NSString *)eventName attributes:(NSDictionary *)attributes {
+    NSMutableDictionary *eventAttributes = attributes.mutableCopy;
+    eventAttributes[MMEEventKeyCreated] = [MMEDate.iso8601DateFormatter stringFromDate:eventDate];
+    eventAttributes[MMEEventKeyEvent] = eventName;
+
+    return [MMEEvent eventWithAttributes:eventAttributes];
+}
+
++ (instancetype)eventWithName:(NSString *)eventName attributes:(NSDictionary *)attributes {
+    return [MMEEvent eventWithDate:MMEDate.date name:eventName attributes:attributes];
 }
 
 + (instancetype)telemetryMetricsEventWithDateString:(NSString *)dateString attributes:(NSDictionary *)attributes {
-    MMEEvent *telemetryMetrics = [[MMEEvent alloc] init];
-    telemetryMetrics.name = MMEEventTypeTelemetryMetrics;
```

Comment:
I think the main use case for the event `name` being added to the events in this way was for debug logging. It allows for some easy access to the event name without having to go through `attributes`.

Reply:
I see you've added a return value from `attributes` further down. Cool! 👍